packages/plexus/README.md

```diff
@@ -1,13 +1,914 @@
+  label?: React.ReactNode;
+  data?: T;
+};
+```
+
+#### `TLayoutVertex`
+
+The underlying vertex data with layout information for a given vertex.
+
+```tsx
+type TLayoutVertex<T = TExtension> = {
+  vertex: TVertex<T>;
+  height: number;
+  left: number;
+  top: number;
+  width: number;
+};
+```
+
+#### `TLayoutEdge`
+
+The combination of the underlying edge data (for a single edge) and the path information for a given edge.
+
+```tsx
+type TLayoutEdge<T = TExtension> = {
+  edge: TEdge<T>;
+  pathPoints: [number, number][];
+};
+```
+
+#### `TLayoutGraph`
+
+Indicates the size and scale of the full graph after it's been laid out.
+
+```tsx
+type TLayoutGraph = {
+  height: number;
+  scale: number;
+  width: number;
+};
+```
+
+#### `ELayoutPhase`
+
+The various phases of the graph layout process.
+
+```tsx
+enum ELayoutPhase {
+  NoData = 'NoData',
+  CalcSizes = 'CalcSizes',
+  CalcPositions = 'CalcPositions',
+  CalcEdges = 'CalcEdges',
+  Done = 'Done',
+}
+```
+
+#### `TRendererUtils`
+
+These utils are made available to the props factories (`setOn...`) and the render fields of the layer configuration objects.
+
+| Field | Type and description |
+| :-- | :-- |
+| getLocalId | `(name: string) => string` |
+|  | Takes in a string and prefixes it to scope the string to the current plexus graph. This effectively allows for IDs that are unique within the document given the `name` parameter is unique within a graph.<br>&nbsp; |
+| getZoomTransform | `() => ZoomTransform` |
+|  | Returns the current D3 zoom transform. See <https://github.com/d3/d3-zoom#zoom-transforms> for details.<br>&nbsp; |
+
+#### `TExposedGraphState`
+
+This type gives access to the graph's current state, such as the current phase or the layout vertices. This is available to the container-level prop factories and the render field for `TDefEntry` elements of SVG layer configurations.
+
+| Field | Type and description |
+| :-- | :-- |
+| edges | `TEdge[]` |
+|  | The user provided edge data underlying the edges in the graph.<br>&nbsp; |
+| layoutEdges | `TLayoutEdge[] \| null` |
+|  | The edge data and the layout data. This is `null` if the layout is not yet generated.<br>&nbsp; |
+| layoutGraph | `TLayoutGraph \| null` |
+|  | The dimensions of the graph. `null` if the layout is not yet generated.<br>&nbsp; |
+| layoutPhase | `ELayoutPhase` |
+|  | The current phase of the graph.<br>&nbsp; |
+| layoutVertices | `TLayoutVertex[] \| null` |
+|  | The vertex data and the layout data. This is `null` if the layout is not yet generated.<br>&nbsp; |
+| renderUtils | `TRendererUtils` |
+|  | Utils for converting an ID local to the graph to globally unique in the document and fetching the current zoom transform.<br>&nbsp; |
+| vertices | `TVertex[]` |
+|  | The user provided vertex data underlying the nodes in the graph.<br>&nbsp; |
+| zoomTransform | `ZoomTransform` |
+|  | The current zoom transform on the graph.<br>&nbsp; |
+
+#### `TContainerPropsSetter`
```

Comment:
Thanks, will resolve.

Reply:
This is not used for `setOnNode` or `setOnEdge`. Those have a different signature so they don't re-render with every change to the zoom transform.

And, because it's used on `setOnEntry` it doesn't only apply to containers :/

What do you think of `TGraphStatePropsSetter` ?

(I think this change should be done on one of the layer PRs so I don't get fairly sizable merge conflicts.)

Reply:
That is a tough one to name. `TGraphStatePropsSetter` seems a bit better. Maybe `TMakePropsFromGraphState`?